File: backend/infrastructure/background.py
"""Registry for fire-and-forget background tasks.

`asyncio.create_task()` alone is not enough for work that nobody awaits: the
event loop keeps only a *weak* reference to a running task, so a task with no
other referent can be garbage-collected mid-flight, and an exception raised
inside it is never observed.

`spawn_background()` holds a strong reference until the task finishes, logs any
exception it raised, and lets shutdown wait for outstanding work.
"""

import asyncio
import logging
from collections.abc import Coroutine
from typing import Any

logger = logging.getLogger("BackgroundTasks")

# Strong references to in-flight tasks, so the GC cannot collect them early.
_background_tasks: set[asyncio.Task] = set()


def _log_task_exception(task: asyncio.Task) -> None:
    """Surface an exception from a task nobody awaits."""
    if task.cancelled():
        return
    exc = task.exception()
    if exc is not None:
        logger.error(f"Background task '{task.get_name()}' failed: {exc}", exc_info=exc)


def spawn_background(coro: Coroutine[Any, Any, Any], *, name: str) -> asyncio.Task:
    """Start a fire-and-forget task that stays referenced and logs its failures.

    Args:
        coro: Coroutine to run.
        name: Task name, used in logs. Include identifying context
            (e.g. ``f"trigger_agent_responses:room={room_id}"``).

    Returns:
        The created task. Callers may ignore it — the registry keeps it alive.
    """
    task = asyncio.create_task(coro, name=name)
    _background_tasks.add(task)
    task.add_done_callback(_background_tasks.discard)
    task.add_done_callback(_log_task_exception)
    return task


async def run_uninterruptible(coro: Coroutine[Any, Any, Any]) -> Any:
    """Run ``coro`` to completion even if the caller is cancelled, then re-raise.

    Use for a write that must not be torn in half -- persisting an agent
    response the user already watched stream in, say. Orchestration tasks are
    cancelled routinely (a new player action interrupts the previous turn), and
    a cancellation landing between the INSERT and the COMMIT loses the message.

    Note this is deliberately *not* a bare ``asyncio.shield``. Shield re-raises
    CancelledError in the caller immediately and leaves the shielded coroutine
    running loose; the enclosing ``background_session()`` would then close the
    session out from under the still-running write. Here the cancellation is
    absorbed until the write has actually landed, and only then re-raised, so
    the caller's teardown never overtakes it.
    """
    task = asyncio.ensure_future(coro)
    cancelled = False

    while not task.done():
        try:
            await asyncio.shield(task)
        except asyncio.CancelledError:
            if task.cancelled():
                raise
            cancelled = True  # ours, not the task's -- keep waiting
        except Exception:
            break  # the task's own failure; re-raised by task.result() below

    if cancelled:
        # The write landed. Now honour the cancellation the caller asked for.
        raise asyncio.CancelledError()

    return task.result()


def pending_background_tasks() -> set[asyncio.Task]:
    """Return the tasks that have not finished yet (for tests and diagnostics)."""
    return {task for task in _background_tasks if not task.done()}


async def drain_background_tasks(timeout: float = 10.0) -> None:
    """Wait for outstanding background tasks, cancelling whatever overruns.

    Called on application shutdown so in-flight agent turns get a chance to
    finish writing before the event loop goes away.
    """
    pending = pending_background_tasks()
    if not pending:
        return

    logger.info(f"Waiting up to {timeout:.0f}s for {len(pending)} background task(s)")
    _done, still_running = await asyncio.wait(pending, timeout=timeout)

    if still_running:
        names = ", ".join(sorted(task.get_name() for task in still_running))
        logger.warning(f"Cancelling {len(still_running)} background task(s) that overran shutdown: {names}")
        for task in still_running:
            task.cancel()
        await asyncio.gather(*still_running, return_exceptions=True)

File: backend/src/infrastructure/background.ts
// Registry for fire-and-forget background work. Two reasons it exists: an
// unobserved rejection is a process-level `unhandledRejection`, fatal under
// Bun's default; and shutdown has to wait for in-flight turns to finish.

import { getLogger } from './logging/logger'

const logger = getLogger('BackgroundTasks')

interface BackgroundTask {
  readonly name: string
  readonly promise: Promise<void>
  readonly abort: AbortController
}

const tasks = new Set<BackgroundTask>()

export interface SpawnOptions {
  /** Task name for logs. Include context: `enter_world:world=12`, not `turn`. */
  name: string
}

/**
 * Start fire-and-forget work that stays registered and logs its failures.
 * `body`'s `AbortSignal` fires only if the task overruns shutdown, and the
 * returned promise never rejects.
 */
export function spawnBackground(
  body: (signal: AbortSignal) => Promise<unknown>,
  { name }: SpawnOptions,
): Promise<void> {
  const abort = new AbortController()

  // Deferred so the cleanup closure can capture `task`: the body may run
  // synchronously up to its first await.
  let settle: () => void = () => {}
  const promise = new Promise<void>((resolve) => {
    settle = resolve
  })

  const task: BackgroundTask = { name, promise, abort }
  tasks.add(task)

  void (async () => {
    try {
      await body(abort.signal)
    } catch (error) {
      if (abort.signal.aborted) {
        logger.warning(`Background task '${name}' was cancelled at shutdown`)
      } else {
        logger.exception(`Background task '${name}' failed`, error)
      }
    } finally {
      tasks.delete(task)
      settle()
    }
  })()

  return promise
}

/** The tasks that have not finished yet (for tests and diagnostics). */
export function pendingBackgroundTasks(): { name: string }[] {
  return [...tasks].map(({ name }) => ({ name }))
}

// Called on shutdown so an in-flight turn can finish writing before the
// database handle closes; whatever overruns is aborted.
export async function drainBackgroundTasks(timeoutMs = 10_000): Promise<void> {
  if (tasks.size === 0) return

  const inFlight = [...tasks]
  logger.info(`Waiting up to ${Math.round(timeoutMs / 1000)}s for ${inFlight.length} background task(s)`)

  let timer: ReturnType<typeof setTimeout> | undefined
  const deadline = new Promise<'timeout'>((resolve) => {
    timer = setTimeout(() => resolve('timeout'), timeoutMs)
  })

  const all = Promise.all(inFlight.map((task) => task.promise)).then(() => 'done' as const)
  const outcome = await Promise.race([all, deadline])
  if (timer !== undefined) clearTimeout(timer)

  if (outcome === 'timeout') {
    const overran = [...tasks]
    if (overran.length > 0) {
      const names = overran
        .map((task) => task.name)
        .sort()
        .join(', ')
      logger.warning(`Cancelling ${overran.length} background task(s) that overran shutdown: ${names}`)
      for (const task of overran) task.abort.abort()
      await Promise.all(overran.map((task) => task.promise))
    }
  }
}

/** Drop all registrations. Tests only — it does not abort anything. */
export function resetBackgroundTasks(): void {
  tasks.clear()
}
